File: src/main/java/DP/MaxSubArray.java
package DP;

public class MaxSubArray {

    public int maxSubArray(int[] nums) {
        int len = nums.length;
        // 定义dp及其下标含义,表示最大子序和
        int[] dp = new int[len];
        // 确定递推公式并初始化
        dp[0] = nums[0];
        int result = nums[0];
        for (int i = 1; i < len; i++) {
            dp[i] = Math.max(dp[i - 1] + nums[i], nums[i]);
            if (dp[i] > result) {
                result = dp[i];
            }
        }
        return result;

    }
}

File: src/main/java/greedyalgorithm/MaxSubArray.kt
package greedyalgorithm

class MaxSubArray {
    fun maxSubArray(nums: IntArray): Int {
        // 贪心算法
        var result = Int.MIN_VALUE
        if (nums.size==1){
            return nums[0]
        }
        var count = 0
        var i = 0
        while (i<nums.size){
            count+=nums[i]
            if (result<count){
                result = count
            }
            if (count<0){
                count =0
            }
            i++
        }
        return result
    }
}